### PR TITLE
[GOBBLIN-772]Implement Schema Comparison Strategy during Disctp

### PR DESCRIPTION
**Pull request by ZihanLi58**
Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [x] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-XXX] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-772

### Description
- [x] Here are some details about my PR, including screenshots (if applicable):
We need a schema comparison strategy to make sure the real schema and the expected schema have compatibility.
- Add an interface of schema check strategy.
- Add configuration using which user can specify the schema check strategy
- Add default schema check strategy. The default schema check strategy will check if the two schemas have matching field names and types.
- The file org.apache.gobblin.restli.throttling.permits.snapshot.json changed because the file gobblin-github/gobblin-restli/gobblin-throttling-service/gobblin-throttling-service-api/src/main/pegasus/org/apache/gobblin/restli/throttling/PermitAllocation.pdsc changed in ticket[GOBBLIN-724]. And after a full build, this change will automatically shows up.

### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
- Unit test: Make the real schema and expected schema have different name or type and make sure the method return false. And make they have matching name and type but not the same doc and make sure the method return true.
- Integration test: Run end to end test on Hadoop to make sure the SchemaCheckedCopySource works well with the strategy.

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

**Comment by ZihanLi58** [comment]
@ibuenros @autumnust Can you take a look at this code?

File: gobblin-restli/gobblin-throttling-service/gobblin-throttling-service-api/src/main/snapshot/org.apache.gobblin.restli.throttling.permits.snapshot.json [hunk 1]
```diff
@@ -17,6 +17,18 @@
       "type" : "long",
       "doc" : "Client should not try to acquire permits before this delay has passed.",
       "optional" : true
+    }, {
```
Comment (autumnust) [inline gobblin-restli/gobblin-throttling-service/gobblin-throttling-service-api/src/main/snapshot/org.apache.gobblin.restli.throttling.permits.snapshot.json]: Is this changes from other PR? Can you rebase the newest changes so that it won't be included?

File: gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/extractor/FileAwareInputStreamExtractorWithCheckSchema.java [hunk 2]
```diff
@@ -39,32 +42,109 @@
...
+  protected FileAwareInputStream buildStream(FileSystem fsFromFile) throws DataRecordException, IOException {
+    if (!schemaChecking(fsFromFile)) {
       throw new DataRecordException("Schema does not match the expected schema");
     }
     return super.buildStream(fsFromFile);
   }
 
-  protected boolean schemaChecking(FileSystem fsFromFile)
-      throws IOException {
+  protected boolean schemaChecking(FileSystem fsFromFile) throws IOException {
```
Comment (autumnust) [inline gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/extractor/FileAwareInputStreamExtractorWithCheckSchema.java]: What is the scope of this schema checking ? Is this verifying all fields inside `toValidateSchema` is appearing in expected schema ? Please add javadoc on this.

File: gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/extractor/FileAwareInputStreamExtractorWithCheckSchema.java [hunk 3]
```diff
@@ -39,32 +42,109 @@
...
+              return false;
+            } else {
+              if (!compare(toValidateField.schema(), expectedFiled.schema())) {
+                return false;
+              }
+            }
+          }
+          return true;
+        }
+        case UNION: {
```
Comment (autumnust) [inline gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/extractor/FileAwareInputStreamExtractorWithCheckSchema.java]: Just want to confirm, do you care about the ordering of fields ?

File: gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/extractor/FileAwareInputStreamExtractorWithCheckSchema.java [hunk 4]
```diff
@@ -39,32 +42,109 @@
...
+    DataFileReader<GenericRecord> dataFileReader =
+        new DataFileReader(new FsInput(this.file.getFileStatus().getPath(), fsFromFile), datumReader);
     Schema schema = dataFileReader.getSchema();
-    return schema.toString().equals(this.state.getProp(ConfigurationKeys.COPY_EXPECTED_SCHEMA));
+    Schema expectedSchema = new Schema.Parser().parse(this.state.getProp(ConfigurationKeys.COPY_EXPECTED_SCHEMA));
+
+    return compare(schema, expectedSchema);
+  }
+
+  private boolean compare(Schema toValidate, Schema expected) {
```
Comment (autumnust) [inline gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/extractor/FileAwareInputStreamExtractorWithCheckSchema.java]: Depending on the scope of your schema checking, this method can be reused for other purposes. May be make it static method or put in some utilities classes?

File: gobblin-restli/gobblin-throttling-service/gobblin-throttling-service-api/src/main/snapshot/org.apache.gobblin.restli.throttling.permits.snapshot.json (hunk 1 above)
Comment (ZihanLi58) [reply gobblin-restli/gobblin-throttling-service/gobblin-throttling-service-api/src/main/snapshot/org.apache.gobblin.restli.throttling.permits.snapshot.json]: Every time I build the project, this change will be automatically made.

File: gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/extractor/FileAwareInputStreamExtractorWithCheckSchema.java (hunk 3 above)
Comment (ZihanLi58) [reply gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/extractor/FileAwareInputStreamExtractorWithCheckSchema.java]: No, I just want to make sure they have matching field.

**Comment by autumnust** [comment]
@ZihanLi58 Travis is failing

File: gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/extractor/FileAwareInputStreamExtractorWithCheckSchema.java [hunk 5]
```diff
@@ -39,32 +43,38 @@
  * check if the schema matches the expected schema. If not it will abort the job.
  */
 
-public class FileAwareInputStreamExtractorWithCheckSchema extends FileAwareInputStreamExtractor{
+public class FileAwareInputStreamExtractorWithCheckSchema extends FileAwareInputStreamExtractor {
```
Comment (jhsenjaliya) [inline gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/extractor/FileAwareInputStreamExtractorWithCheckSchema.java]: just a comment on this class: i understand that this class already exists and its adding proper field level schema check, but can schemaCheck not be part of `FileAwareInputStreamExtractor` itself instead of introducing another class for schemaCheck functionality, since that is the only difference with this class which probably can be controlled by a schemaCheck flag. may be its a separate PR.

File: gobblin-data-management/src/main/java/org/apache/gobblin/util/schema_check/AvroSchemaCheckStrategy.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+     * Use the configuration to create a schema check strategy. If it's not found, return null.
+     * @param state
+     * @return
+     */
+    public static AvroSchemaCheckStrategy create(WorkUnitState state)
+    {
+      try {
+        return (AvroSchemaCheckStrategy) Class.forName(state.getProp(ConfigurationKeys.AVRO_SCHEMA_CHECK_STRATEGY, ConfigurationKeys.AVRO_SCHEMA_CHECK_STRATEGY_DEFAULT)).newInstance();
+      } catch (Exception e)
+      {
```
Comment (autumnust) [inline gobblin-data-management/src/main/java/org/apache/gobblin/util/schema_check/AvroSchemaCheckStrategy.java]: nitpick: usually we would have `{` in the same line of `catch`

File: gobblin-data-management/src/main/java/org/apache/gobblin/util/schema_check/AvroSchemaCheckStrategy.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+        return null;
```
Comment (autumnust) [inline gobblin-data-management/src/main/java/org/apache/gobblin/util/schema_check/AvroSchemaCheckStrategy.java]: Exception caught but not propagated/logged will be a silent failure which is usually not what we wanted. If there's nothing (recovery type of thing) we could do after catching the exception, let's just log the exception as a starting point

File: gobblin-restli/gobblin-throttling-service/gobblin-throttling-service-api/src/main/snapshot/org.apache.gobblin.restli.throttling.permits.snapshot.json (hunk 1 above)
Comment (autumnust) [reply gobblin-restli/gobblin-throttling-service/gobblin-throttling-service-api/src/main/snapshot/org.apache.gobblin.restli.throttling.permits.snapshot.json]: Got the point. Just mention in the description that why this changes (which is unrelated to the PR itself) shows up as part of code change.

File: gobblin-api/src/main/java/org/apache/gobblin/configuration/ConfigurationKeys.java [hunk 8]
```diff
@@ -936,4 +936,10 @@
   public static final String GIT_MONITOR_SSH_KNOWN_HOSTS = "knownHosts";
   public static final String GIT_MONITOR_SSH_KNOWN_HOSTS_FILE = "knownHostsFile";
   public static final String GIT_MONITOR_JSCH_LOGGER_ENABLED = "isJschLoggerEnabled";
+
+  /**
+   * Configuration related to data derived
```
Comment (ibuenros) [inline gobblin-api/src/main/java/org/apache/gobblin/configuration/ConfigurationKeys.java]: Please adjust the comment so it makes sense in Gobblin itself. This is leaking LinkedIn implementation details.

File: gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/extractor/FileAwareInputStreamExtractorWithCheckSchema.java (hunk 5 above)
Comment (ibuenros) [reply gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/extractor/FileAwareInputStreamExtractorWithCheckSchema.java]: @jhsenjaliya the reason we did this in a different class is because it is leaking Avro details. CopySource and the associated extractor are able to copy arbitrary files simply as byte streams. This particular extension requires the copied files to be Avro.

In general, I think that in a byte copier any functionality that assumes a schema associated with those bytes is pollution of the logic. What do you think?

File: gobblin-data-management/src/test/java/org/apache/gobblin/data/management/util/AvroSchemaCheckStrategyForTest.java [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+package org.apache.gobblin.data.management.util;
+
+import java.util.HashSet;
+import java.util.Set;
+import org.apache.avro.AvroRuntimeException;
+import org.apache.avro.Schema;
+import org.apache.gobblin.util.schema_check.AvroSchemaCheckStrategy;
+
+
+public class AvroSchemaCheckStrategyForTest implements AvroSchemaCheckStrategy {
```
Comment (ibuenros) [inline gobblin-data-management/src/test/java/org/apache/gobblin/data/management/util/AvroSchemaCheckStrategyForTest.java]: Is there a reason this needs to be only for testing? Is it robust enough to be made the default implementation in Gobblin?

**Comment by ibuenros** [comment]
+1

File: gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/extractor/FileAwareInputStreamExtractorWithCheckSchema.java (hunk 5 above)
Comment (jhsenjaliya) [reply gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/extractor/FileAwareInputStreamExtractorWithCheckSchema.java]: ok, sure. Thanks
